src/JobOrderer.ts

```diff
@@ -0,0 +1,138 @@
+import { assert } from 'console'
+import { Job } from './Job'
+
+/**
+ * Manages a set of jobs that have to be run.
+ */
+export class JobOrderer {
+	private sources: Set<Job> = new Set()
+	private nonSources: Set<Job> = new Set()
+	private inProgress: Set<Job> = new Set()
+	private jobToDependents: Map<Job, Set<Job>> = new Map()
+
+	/**
+	 * Create the JobOrderer and inform it of the jobs to manage.
+	 *
+	 * @throws When one of the jobs has a prerequisite that is not one of the jobs, or is itself.
+	 * @param jobs - The jobs to manage. They must have correctly populated prerequisites fields.
+	 */
+	constructor(jobs: Job[]) {
+		// Initialize dependents map.
+		jobs.forEach((job) => this.jobToDependents.set(job, new Set()))
+
+		// Populate dependents map.
+		jobs.forEach((job) => {
+			job.incompletePrerequisites.forEach((prerequisite) => {
+				if (prerequisite == job) {
+					throw `Circular dependency: job ${job} has itself as a preqrequisite.`
```

Comment:
Which of those two would you recommend?

Reply:
For now I'm leaning towards the former, since that's easy and everything still works. If needed, we can go with the industrial-strength option later :)